internal/saucecloud/cloud.go

```diff
@@ -419,65 +420,171 @@ func (r *CloudRunner) remoteArchiveProject(project interface{}, folder string, s
+	if need {
+		nodeModulesURI, err := r.handleNodeModules(tempDir, projectDir, matcher, dryRun)
+		if err != nil {
+			return "", nil, err
+		}
+		if nodeModulesURI != "" {
+			uris[nodeModulesUpload] = nodeModulesURI
+		}
+	}
+
+	var sortedURIs []string
+	for _, t := range []uploadType{runnerConfigUpload, nodeModulesUpload, otherAppsUpload} {
+		if uri, ok := uris[t]; ok {
+			sortedURIs = append(sortedURIs, uri)
+		}
+	}
 
-	modZip, err := zip.ArchiveNodeModules(tempDir, folder, matcher, r.NPMDependencies)
+	return uris[projectUpload], sortedURIs, nil
+}
+
+// collectFiles retrieves all relevant files in the project directory, excluding "node_modules".
+func collectFiles(dir string) ([]string, error) {
+	var files []string
+	contents, err := os.ReadDir(dir)
 	if err != nil {
-		return
+		return nil, fmt.Errorf("failed to read project directory: %w", err)
 	}
-	if modZip != "" {
-		archives[nodeModulesUpload] = modZip
+
+	for _, file := range contents {
+		if file.Name() != "node_modules" {
+			files = append(files, filepath.Join(dir, file.Name()))
+		}
 	}
+	return files, nil
+}
 
-	configZip, err := zip.ArchiveRunnerConfig(project, tempDir)
+func (r *CloudRunner) createArchives(tempDir, projectDir string, project interface{}, files []string, matcher sauceignore.Matcher) (map[uploadType]string, error) {
+	archives := make(map[uploadType]string)
+
+	projectArchive, err := zip.ArchiveFiles("app", tempDir, projectDir, files, matcher)
 	if err != nil {
-		return
+		return nil, fmt.Errorf("failed to archive project files: %w", err)
 	}
-	archives[runnerConfigUpload] = configZip
+	archives[projectUpload] = projectArchive
 
-	var uris = map[uploadType]string{}
-	for k, v := range archives {
-		uri, err := r.uploadArchive(storage.FileInfo{Name: v}, k, dryRun)
+	configArchive, err := zip.ArchiveRunnerConfig(project, tempDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to archive runner configuration: %w", err)
+	}
+	archives[runnerConfigUpload] = configArchive
+
+	return archives, nil
+}
+
+// handleNodeModules archives the node_modules directory and uploads it to remote storage.
+// If tagging is enabled and a tagged version of node_modules already exists in storage,
+// it returns the URI of the existing archive.
+// Otherwise, it creates a new archive, uploads it and returns the storage ID.
+func (r *CloudRunner) handleNodeModules(tempDir, projectDir string, matcher sauceignore.Matcher, dryRun bool) (string, error) {
+	var tags []string
+
+	if taggableModules(projectDir, r.NPMDependencies) {
+		tag, err := hashio.HashContent(filepath.Join(projectDir, "package-lock.json"), r.NPMDependencies...)
 		if err != nil {
-			return "", []string{}, err
+			return "", err
+		}
+		tags = append(tags, tag)
+
+		log.Info().Msgf("Searching remote node_modules archive by tag %s", tag)
+		existingURI := r.findTaggedArchives(tag)
+		if existingURI != "" {
+			log.Info().Msgf("Skipping archive and upload node_modules, use %s", existingURI)
```

Comment:
Consistency with https://github.com/saucelabs/saucectl/blob/37f05d95eda70674bf443b1675f2d200560dd684/internal/saucecloud/cloud.go#L731-L731
```suggestion
			log.Info().Msgf("Skipping upload, using %s", existingURI)
```